### PR TITLE
Submissions mock

### PR DESCRIPTION
**Pull request by Tthomas63**
* Merge `Dockerize` branch
 * Initial Submission Model Changes
 * Initial Riot Tag from datasets
 * Various celery fixes
 * Modifying RiotJS
 * Submissions work:
    - Files upload
    - Table Reflects Attributes
    - Can download from table
 * Basic Scoring Task implemented
 * Scores Submission when submission is saved and does not have a score

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM python:3
+ENV PYTHONUNBUFFERED 1
+RUN mkdir /code
+WORKDIR /code
+ADD requirements.txt /code/
+RUN pip install -r requirements.txt
+ADD . /code/
```
Comment (ckcollab) [inline Dockerfile]: use volume!

File: Dockerfile.flower [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+# Dockerfile based on the official Celery
+# https://github.com/docker-library/celery
+
+# Prepare app folders & cmd
+RUN pip install flower
+
+CMD ["flower", \
+     "--broker=amqp://admin:admin@rabbitmq:5672/comps", \
+     "--broker_api=http://admin:admin@rabbitmq:15672/api/", \
```
Comment (ckcollab) [inline Dockerfile.flower]: these are hardcoded here... see how we did it on v1.5, you don't need this, there's an image

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    depends_on:
+      - rabbitmq
+      - db
+    environment:
+      - "DJANGO_SETTINGS_MODULE=src.settings.base"
+    env_file: .env
+    volumes:
+      - .:/code
+    command: celery -A src.comp_worker worker --loglevel=debug
+  flower:
```
Comment (ckcollab) [inline docker-compose.yml]: please break this whole file up between each section with long comments, like:

```
# ---------------------------------------------------
# Message queue
# ---------------------------------------------------
.. stuff about rabbit
.. stuff about flower
```

File: src/apps/competitions/migrations/0008_auto_20180130_1104.py [hunk 4]
```diff
@@ -1,7 +1,7 @@
 # Generated by Django 2.0.1 on 2018-01-30 11:04
 
 from django.db import migrations, models
-import utils.data
+import src.utils.data
```
Comment (ckcollab) [inline src/apps/competitions/migrations/0008_auto_20180130_1104.py]: remove `src`

File: src/apps/api/views/competitions.py [hunk 5]
```diff
@@ -29,6 +29,6 @@ class PhaseViewSet(ReadOnlyModelViewSet):
     serializer_class = serializers.PhaseSerializer
 
 
-class SubmissionViewSet(ReadOnlyModelViewSet):
+class SubmissionViewSet(ModelViewSet):
```
Comment (ckcollab) [inline src/apps/api/views/competitions.py]: Can you create a submission for a competition you are not a participant of?

File: src/apps/competitions/models.py [hunk 6]
```diff
@@ -36,6 +38,27 @@ class Submission(models.Model):
...
+
+    # uber experimental
+    track = models.IntegerField(default=1)
+
+    def save(self, force_insert=False, force_update=False, using=None,
+             update_fields=None):
+
+        super(Submission, self).save()
+
+        if not self.score:
```
Comment (ckcollab) [inline src/apps/competitions/models.py]: good stuff

File: src/apps/competitions/tasks.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
+from celery import shared_task
+
+from competitions import models
+
+
+@shared_task
```
Comment (ckcollab) [inline src/apps/competitions/tasks.py]: why "shared task?"

File: src/apps/api/views/competitions.py (hunk 5 above)
Comment (Tthomas63) [reply src/apps/api/views/competitions.py]: Currently yes. I've considered this and it comes down to either assuming the phase we're given ( Which won't be from the user ), is for the current competition, or explicitly defining the competition and the phase, and double checking them against or each other or some such which seems unnecessary

File: src/apps/competitions/migrations/0008_auto_20180130_1104.py (hunk 4 above)
Comment (Tthomas63) [reply src/apps/competitions/migrations/0008_auto_20180130_1104.py]: Needed because of dockerize

Comment (Tthomas63) [reply src/apps/competitions/migrations/0008_auto_20180130_1104.py]: We should change this in that branch... Or change how we're doing things there

File: src/apps/competitions/tasks.py (hunk 7 above)
Comment (Tthomas63) [reply src/apps/competitions/tasks.py]: This is what I am familiar with from Chahub and just what I've used. I'm honestly not sure about task vs shared_task.

File: src/apps/api/views/competitions.py (hunk 5 above)
Comment (ckcollab) [reply src/apps/api/views/competitions.py]: Yeah, phase is tied to competition so no need to specify competition.

I would add a check to see if user is in the competition 

```
if not phase.competition.participants.filter(user=request.user).exists():
    raise PermissionError("something")
```

File: src/apps/competitions/migrations/0008_auto_20180130_1104.py (hunk 4 above)
Comment (ckcollab) [reply src/apps/competitions/migrations/0008_auto_20180130_1104.py]: This def shouldn't be needed, src isn't even a module so it shouldn't even work... confused on this one

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,6 @@
+FROM python:3
+ENV PYTHONUNBUFFERED 1
+RUN mkdir /code
+WORKDIR /code
+ADD requirements.txt /code/
```
Comment (ckcollab) [inline Dockerfile]: not needed?

Comment (ckcollab) [reply Dockerfile]: (with volumes)

File: Dockerfile.celery [hunk 9]
```diff
@@ -0,0 +1,8 @@
+FROM python:3
+ENV PYTHONUNBUFFERED 1
+ENV C_FORCE_ROOT True
+RUN mkdir /code
+WORKDIR /code
+ADD requirements.txt /code/
```
Comment (ckcollab) [inline Dockerfile.celery]: not needed

File: manage.py [hunk 10]
```diff
@@ -8,6 +8,7 @@
     settings_module = os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'settings.base')
 
     if 'runserver' in sys.argv:
+        print("npm-watching")
```
Comment (ckcollab) [inline manage.py]: pull?

File: requirements.txt [hunk 11]
```diff
@@ -19,3 +19,13 @@ asgi_redis==1.4.3
 drf-writable-nested==0.4.1
 django-extra-fields==0.9
 pillow==5.0.0
+# New Celery Requirements
```
Comment (ckcollab) [inline requirements.txt]: comment not necessary

File: src/apps/api/views/competitions.py (hunk 5 above)
Comment (Tthomas63) [reply src/apps/api/views/competitions.py]: Alright, we'll keep this in mind

File: docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,79 @@
+version: '3'
+volumes:
+  dbvolume:
```
Comment (ckcollab) [inline docker-compose.yml]: `dbvolume` we should rename to `persistant_dir` and we should put RabbitMQ and Postgres storage in there

File: requirements.txt [hunk 13]
```diff
@@ -19,3 +19,12 @@ asgi_redis==1.4.3
 drf-writable-nested==0.4.1
 django-extra-fields==0.9
 pillow==5.0.0
+celery==3.1.23
```
Comment (ckcollab) [inline requirements.txt]: eeeep!!! Use Celery 4! Django-celery isn't even a thing any more I don't think?

File: src/apps/competitions/tasks.py (hunk 7 above)
Comment (ckcollab) [reply src/apps/competitions/tasks.py]: yeah let's just use task

File: src/comp_worker.py [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'settings.base')
+
+app = Celery('comps')
+
+from django.conf import settings  # noqa
+#
+app.config_from_object('django.conf:settings')
+app.autodiscover_tasks(lambda: settings.INSTALLED_APPS)
+
+# @app.task(bind=True)
```
Comment (ckcollab) [inline src/comp_worker.py]: pull this stuff?

File: src/settings/base.py [hunk 15]
```diff
@@ -174,6 +188,17 @@
     # Allows us to use with django-oauth-toolkit on localhost sans https
     SESSION_COOKIE_SECURE = False
 
+# ============================================================================
+# Celery
+# ============================================================================
+
+import djcelery
```
Comment (ckcollab) [inline src/settings/base.py]: pull this

File: src/static/js/ours/client.js [hunk 16]
```diff
@@ -31,13 +31,64 @@ CODALAB.api = {
...
+                * POST to server
+
+            Remote storage:
+                * POST to server
+                * Server returns SAS URL
+                * PUT to SAS URL
+                * POST to mark upload as done, so un-finished uploads can be pruned later
+
+        */
+        return $.ajax({
```
Comment (ckcollab) [inline src/static/js/ours/client.js]: can we abstract this nice function out? I think it's basically exactly the same for dataset/submission?

File: src/static/riot/admin/admin_submissions.tag [hunk 17]
```diff
@@ -0,0 +1,202 @@
+<!-- Field class on initial definition to keep Semantic UI styling -->
+<field class="field">
```
Comment (ckcollab) [inline src/static/riot/admin/admin_submissions.tag]: is this a repeat definition of `<field>`?

File: src/static/riot/admin/admin_submissions.tag [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+        <div class="actions">
+            <div class="ui cancel button">I've copied down this key</div>
+        </div>
+    </div>
+
+    <script>
+        // --------------------------------------------------------------------
+        // Initializers
+        var self = this
+        self.comp_pk = 1
```
Comment (ckcollab) [inline src/static/riot/admin/admin_submissions.tag]: set to undefined here just in case

File: src/static/riot/admin/admin_submissions.tag [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+            </form>
+        </div>
+        <div class="actions">
+            <input type="submit" class="ui button" form="submission_form" value="Save"/>
+            <div class="ui cancel button">Cancel</div>
+        </div>
+    </div>
+
+    <!-- New key modal -->
+    <div id="submission_secret_key_modal" class="ui modal">
```
Comment (ckcollab) [inline src/static/riot/admin/admin_submissions.tag]: can pull this

File: docker-compose.yml [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+      - ./var/postgres:/var/lib/postgresql/data
+  #-----------------------------------------------
+  #   Rabbitmq Service
+  #-----------------------------------------------
+  rabbitmq:
+    image: rabbitmq:3-management
+    container_name: rabbit
+    env_file: .env
+    ports:
+      - "15672:15672"
```
Comment (ckcollab) [inline docker-compose.yml]: remove quotes on all of these ports

File: requirements.txt [hunk 21]
```diff
@@ -19,3 +17,13 @@ asgi_redis==1.4.3
 drf-writable-nested==0.4.1
 django-extra-fields==0.9
 pillow==5.0.0
+# celery==3.1.23
+# django-celery==3.2.1
+# django-celery-beat==1.1.0
+# django-celery-results==1.0.1
+celery
```
Comment (ckcollab) [inline requirements.txt]: version this

File: src/apps/competitions/tasks.py [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+
+from competitions import models
+
+
+@app.task
+def score_submission_lazy(submission_pk):
+    sub_to_score = models.Submission.objects.get(pk=submission_pk)
+    sub_to_score.score = 1
+    sub_to_score.save()
+    return True  # Do we need to do this?
```
Comment (ckcollab) [inline src/apps/competitions/tasks.py]: no

File: requirements.txt (hunk 13 above)
Comment (ckcollab) [reply requirements.txt]: Still not celery 4?

**Pull request merged**